Add best effort toolchain feature docs (#701)
bazelbuild/rules_cc: docs/features.md

# CC Toolchain Features

NOTE: It's possible this document has drifted, please file issues or
submit PRs for any inaccuracies you find

> ## Toolchain features

Reviewer: It would be great in the future to elaborate a bit more on this section. As a user some parts surprise me but maybe I'm not yet well familiar with the vision for rules_cc. I guess that such written vision for the entire rules_cc would be required first.
Right now rules_cc mentions in the main readme that it provides autodetected toolchains, and that there are other toolchains that provide hermetic toolchains.
https://github.com/bazelbuild/rules_cc#using-the-rules_cc-toolchain
A very fundamental use of toolchain features, is to provide the same semantic meaning accross toolchains. As a user, you can use the feature name to communicate what you want and the different toolchains can implement it the best way they can.
In the moment that you say that the feature names are not part of the public API, raises a bit of concerns to me. Because this would mean that every time a feature name changes, another toolchain that is not part of rules_cc would have to be adapted.
Every time that I thought what should be part of rules_cc or not, I always had mix feelings. One one side rules_cc cannot consider every single C++ compiler, compiler version, and build flags. On the otherside, leaving that part to compiler specific toolchain might not be the best, considering that there are some kind of compiler families that they user very similar flags.
In the end with this comment I just wanted to share my thoughts. This is great work and it is clearly better to have it here than having it in code. At least with a readme it might job a bit better intention.
I believe that any improvement can be done afterwards.

Reviewer 2: @limdor I share the same opinion.

In rules_cc, `features` fit on a fence. The communication around makes user believe `features` are not part of the public API. But in practice they are public API because you can't implement a cc_toolchain without using well-known features.

Also, C++ library developer can't decouple their BUILD file from the toolchain implementation. For instance if you want to enable C++20, you either need to believe the toolchain expose a feature for that (e.g `cxx_standard_20`) or you need to branch on `@rules_cc//cc/compiler`'s value to select between `-std=c++20` for gcc/clang or `/std:c++20` for msvc. Either the feature or the compiler name needs to be standardized.

@keith is there a desire from the rules_cc's maintainers to progress into that direction? If we would like to contribute what is the best way? start by writing a design doc?

Author: i wrote the thing about public API since i am trying to include _every_ feature that exists in rules_cc in this doc, many of which are legacy holdovers that I don't think rules_cc wants to support forever. And so i didn't want to make it feel like this doc existing upgraded the support contract in any way. I agree that many features names like `layering_check` are something that have to be considered public api. I don't really want to try to put a label on what is supported in this one, I think if we want to have a hard line on what is public API or not we should actually change the code a bit. I imagine we could make a `features.bzl` that is similar to `action_names.bzl`, and sort them there, then update this doc accordingly

CC toolchains are configured by creating features. Features are
arbitrary strings for enabling or disabling behavior in the toolchain.

Semantically there are 3 types of features:

1. Features whose names are arbitrary, and are only used to carry
   command line flags.

> 2. Features with special names which are markers to bazel / `rules_cc`

Reviewer 3: Aside: Perhaps we should promote "marker feature" to a glossary term we use consistently in various places. I've used `sentinel feature` in the past, but I'm not picky about which we use as long as it's consistent so the term is searchable.

   various compiler / linker flags to enable some behavior.

With all features, even though the feature name may have special meaning
to `rules_cc`, it is still up to your toolchain to provide the correct
compiler flags for your situation.

Depending on how bazel / `rules_cc` read the features, you might need to
define them differently.

In some cases `rules_cc` checks if a feature is *supported*, or it
automatically _enables_ it when it's relevant. In this case that means a
feature with that name is defined in the toolchain. For example:

```bzl
return [
    cc_common.create_cc_toolchain_config_info(
        features = [
            feature(name = "dbg"), # Supported by the toolchain but off by default
        ],
        ...
    ),
]
```

This is separate from if a feature is *enabled*, which either means a
feature is defined in the toolchain _and_ automatically enabled:

```bzl
feature(
    name = "archive_param_file",
    enabled = True,  # Enabled by default when defined
),
```

Or the feature is enabled by a user passing
`--features=archive_param_file`, setting `features =
["archive_param_file"]` or through other mechanisms in the toolchain

> definition (not covered here). This distinction is important for when

Reviewer 3: The Either/Or is split by prose and a code block in a way that makes it hard to see where the clause starts/ends. Consider splitting the boundary with bullets:

````suggestion
This is separate from when a feature is *enabled*, which means one of the following:
* The feature is defined in the toolchain _and_ enabled by default in its definition:
  ```bzl
  feature(
      name = "archive_param_file",
      enabled = True,  # Enabled by default when defined
  )
  ```
* The feature is enabled through another mechanism: `--features=archive_param_file` from the
  command line, as requested by a rule when configuring features, or implicitly through a
  relationship to another rule.

This distinction is important for when
```

linker flags. This is required for use with `LTO`.

This feature must be enabled if supported.

### `symbol_check`

A feature automatically requested by `cc_static_library` that enables
the toolchain to enable optional validation around the symbols in the
produced static library.

This feature should be off by default and is automatically requested by
`cc_static_library`.

### `system_include_paths`

A marker feature indicating that include paths from the `includes`
attribute of a target should be passed with `-isystem` instead of `-I`.
This is still up to the toolchain to configure correctly, but this
affects the toolchain variables the include paths are passed through.
This is expected to be set by users when necessary (hopefully rarely).

This feature takes effect even if the toolchain doesn't define it. There
is no purpose in adding it to your toolchain unless you want to enable
it everywhere.

### `targets_windows`

This is used by `rules_cc` to change the behavior in various places only
when building for Windows. If your toolchain targets Windows this should
be enabled.

This feature must be enabled if targeting Windows.

> ### `treat_warnings_as_errors`

Reviewer 2: I have the impression `treat_warnings_as_errors` may not belong to that list.

In the contrary of all other feature which are recognized by bazel and rules_cc, this one is just a name convention used by default toolchains, no?

Reviewer: This feature to me is a very good example that from implementation point of view does not need to be on rules_cc, but all toolchains need to agree on some semantics.
That is part of what I tried to expresed in here.
https://github.com/bazelbuild/rules_cc/pull/701/changes#r3143602971
If for examples toolchains_llvm decide that the way to treat warnings as error is to use a feature called `warnings_as_errors` but then the msvc toolchain uses `all_warnings_as_errors` then the whole point of toolchain features is defeated.
This is feature is a very clear one that we would probably all like to have an agreed name, but there are others. At some point it would be good if the Bazel C++ community sits together and we agree on some feature names and semantics that all toolchains can provide.

Author: I chose to include this one since it is user facing today, and is supported by 2 toolchains here as well as apple_support

These features should be off by default.

### `generate_linkmap`

This feature is requested based on
[`--objc_generate_linkmap`](https://bazel.build/reference/command-line-reference#flag--objc_generate_linkmap)
and commonly correlates with the `-map` linker flag.

This feature should be off by default.

### `generate_pdb_file`

This feature indicates a Windows `pdb` file should be created when
linking a binary. This must be enabled by the user or the toolchain.

This feature must be enabled if desired.

### `lang_objc`

A marker feature indicating that Objective-C or Objective-C++ is being
built.

This feature should be off by default.

### `layering_check`

Enable validation that a library directly depends on everything it uses.
This is implemented using `clang`'s `modulemap` features. See the
default toolchains for implementation examples. `rules_cc` does not
reference this feature directly, but the name `layering_check` is used
by users to enable this behavior, and disable it for incompatible
targets.

This feature should be off by default and turned on at the project /
target level.

### LTO features

> Bazel / `rules_cc` have a man special features for LTO behavior:

Author: ```suggestion
Bazel / `rules_cc` have many special features for LTO behavior:
```

- `thin_lto` top level feature that is also used by users
- `thin_lto_all_linkstatic_use_shared_nonlto_backends` read the source
- `thin_lto_linkstatic_tests_use_shared_nonlto_backends` read the source
- `no_use_lto_indexing_bitcode_file` read the source
- `use_lto_native_object_directory` read the source

> These feature must be enabled if desired.

Author: ```suggestion
These features must be enabled if desired.
```

# CC Toolchain Features

> This document covers the special strings used by bazel / `rules_cc` to

Reviewer 2: I don't think it is necessary to defines what is a feature here since it will be done in the first paragraph two lines below. It could be shorten to:

> This document covers `rules_cc` well-known C++ toolchain features.

configure cc toolchains behavior. This is useful for flipping specific
behavior on or off for your project, or for writing your own toolchain.

> NOTE: It's possible this document has drifted, please file issues or

Reviewer 2: I think mentioning the reason for the drift would improve clarity with minimal added wording:
> This document may drift from the rules_cc implementation, as it is not kept in sync. Please file issues or submit PRs for any inaccuracies

Author: tbh there isn't a good reason like that, i think once this merges if i submit prs changing this stuff, i will update the docs, and i will review prs asking folks to do that too

submit PRs for any inaccuracies you find

> ## Toolchain features

Reviewer 2: I think it is possible to restructure this paragraph by first introducing the concept of feature, then the different types of features.

I can imagine the following structure:
- Features Basics
  - feature definition
  - feature state (defined and enabled)
  - what enabled feature do: add cli args and modify rules logic
  - how are features enabled
- Features Types
  - Marker Features: enable logic
  - Configuration Features: are enabled by certain configuration state (etc. --compilation_mode=opt)
  - Conventional Features: named not recognized by rules_cc, but convention in the bazel ecosystem (e.g treat_warnings_as_error)

Those types can be used as category to group features in the reference paragraph below

Depending on how bazel / `rules_cc` read the features, you might need to
define them differently.

In some cases `rules_cc` checks if a feature is *supported*, or it
automatically enables it when it's relevant. In this case that means a
feature with that name is defined in the toolchain. For example:

```bzl
return [
    cc_common.create_cc_toolchain_config_info(
        features = [
            feature(name = "dbg"), # Supported but off by default
        ],
        ...
    ),
]
```

This is separate from if a feature is *enabled*, which either means a
feature is defined in the toolchain and automatically enabled:

```bzl
feature(
    name = "archive_param_file",
    enabled = True, # Enabled by default when defined
),
```

Or the feature is enabled by a user passing
`--features=archive_param_file`, or through `implies` in the toolchain
itself (not covered here). This distinction is important for when
`rules_cc` checks if a feature is enabled, without automatically
enabling it. This is common for "marker" features. If `rules_cc` checks
for a feature being enabled, it not existing in the toolchain will be
treated the same as it being disabled. This means you can omit

> features that you do not want to support the behavior of.

Reviewer 4: ```suggestion
treated the same as it being disabled. This means a toolchain that does not want to support a feature can just omit it.
```

In some cases `rules_cc` bases behavior on the presence of a feature,
but doesn't require it to be enabled. This is rare but for the relevant
features below.

NOTE: Feature names aren't really considered public API, and are subject
to change more frequently than the rest of the API (even though their
behavior likely doesn't change often).

NOTE: While this document attempts to cover `rules_cc` behavior, it is
possible for any custom rule to read the features of the toolchain and
change its behavior based on them.

> ## Toolchain variables

Reviewer 2: I don't grasp the purpose of talking about Toolchain variables in this document.

You mention to read the document it is important to be familiar with the concept of variables, but the concept is almost never used below.

I have the impression the main message of the paragraph is "be careful features and variables can have the same name, but they are not the same things". But is it very important here?

If the purpose of the document is to properly document for every marker feature what variable is made available for which action, then variable should be introduced in more details with the concept of args.

But I believe it can be skipped.

In order to read this document, or a CC toolchain, you should be
familiar with the concept of toolchain variables which are separate from
features. One of the ways that `rules_cc` controls toolchain behavior is
by passing values through "variables". These are read by the toolchain
through the configuration such as `expand_if_available` (to see if a a
variable is set) and `iterate_over` (to loop over an array of strings,
such as include paths). These variables often use the same strings as
the feature names, which can be confusing.

For example the `pic` feature (discussed below) is a special _feature_
name that also reads the `pic` variable, which is a special _variable_
name decided on by `rules_cc`. The fact that the same name is used is
not a requirement but is a common pattern to be aware of.

> ## Toolchain actions

Reviewer 2: I think it is more confusing than relevant to talk about action_config in this document:
- it looks orthogonal to features. Removing action_config will not prevent to understand the rest of the document
- action_config is kind of implementation details if we consider rules-based toolchain definition is recommended way to go

This document doesn't cover `action_config`s, but the legacy behavior
(described below) applies to them as well, and the names of the action
configurations are also special.

For example the `strip` action configuration is checked for by name to
determine if `strip` can be run to produce `cc_binary`'s implicit
`binary.stripped` output.

## Legacy features

By default, unless you add the
[`no_legacy_features`](#no_legacy_features) feature to your toolchain,
you will automatically inherit features and action configurations
defined in
[`legacy_features.bzl`](../cc/private/toolchain_config/legacy_features.bzl).
It is recommended that you override all features to avoid this
potentially confusing behavior. You can read that file to see the
current defaults.

If you do not add a `no_legacy_features` feature, any features you add
with the same name as a legacy feature will override the default
behavior, but any that you omit will be added to your toolchain
implicitly.

NOTE: Not all of these features work in the default toolchain, so some
might only be relevant if you are writing your own toolchain.

> ### Legacy features with special names

Reviewer 2: In my understanding, the fact a feature is a legacy feature is orthogonal to the fact a feature is a marker to enable some rules logic, and orthogonal to the fact a feature will be enabled by some bazel flags.

I think it is more confusing than relevant to make the distinction here. A small paragraph to explain that some features are defined by default (defined and enabled?) should be enough.

Then in the long section documenting all features, a mention could be added to inform the reader if the feature is part of the legacy feature set.

#### `sanitize_pwd`

A marker feature indicating the toolchain has sanitized the `PWD` from
the outputs. Otherwise `rules_cc` will set `PWD=/proc/self/cwd` (unless
on macOS) when linking a binary. This is commonly used when
`-fdebug-prefix-map` is supported by the compiler.

This feature must be enabled by default if supported.

#### `set_soname`

A marker feature that causes interface libraries to respect the `soname`
they have. Otherwise `-soname` is passed when creating interface libraries.

This feature must be enabled by default if supported.

#### `serialized_diagnostics_file`

A marker feature for enabling generating a serialized diagnostics file
from the compiler. Commonly used with the `--serialize-diagonostics`
`clang` flag.

This feature should be off by default and requested through `--features`
when desired.

#### `shorten_virtual_includes`

A marker feature that causes virtual include paths generated by
`strip_include_prefix` and friends to use a shorter path. This is useful
on Windows to avoid long path issues.

This feature must be enabled by default if desired.

#### `static_link_cpp_runtimes`

A marker feature used by `rules_cc` to determine if the toolchain

> supports statically

Reviewer 5: missing end of sentence.
